Refactor kube filter tests and consequently fix some highlighted bugs (2/2)

yann-soubeyrand opened the pull request.

yann-soubeyrand commented:
Hi @edsiper,

Can you have a look at this PR please (I've just rebased it on master)? It refactors tests of the kubernetes filter plugin to ease their management. It also adds tests and fixes the bugs it allowed to catch. Two other PR I made (#1265 which enhances exclude annotation to make it work per stream and per container and #1306 which keeps meta extracted from log files when meta can't be obtain form API server and thus fixes a memory leak) are based on this one.

yann-soubeyrand commented:
No idea why the tests are failing. It's like older code is being compiled (slashes aren't escaped in the output :-/).

edsiper commented:
@yann-soubeyrand after your comment I reconsidered the \/ stuff, so dropped that because it will generate more problems than solutions, so code was updated.

Note that your PR is applying many changes and touching 157 files. I know this is adds extra work but let's have one fix/feature per PR.

yann-soubeyrand commented:
> @yann-soubeyrand after your comment I reconsidered the / stuff, so dropped that because it will generate more problems than solutions, so code was updated.

@edsiper what I don't understantd is that, unless I missed something, your revert wasn't part of my PR since I rebased it before you did the change. So I can't explain the tests results :-/

> Note that your PR is applying many changes and touching 157 files. I know this is adds extra work but let's have one fix/feature per PR.

I've split my PR in two parts: the first one (#1420) refactors the test runner to make working with tests easier. The second part (this PR) reafactors tests to have better test harness (tests were missing and some tests were not testing what they were expected to test). I can't split these two PR further because the additional commits fixing things are needed for the tests to pass (the refactor uncovered some bugs which had to be fixed in order for the tests to pass).

So please review #1420 first and once it's merged, I'll rebase this PR (hopefully this will be the last time I'll rebase it).

yann-soubeyrand commented:
@edsiper ping ;-)

yann-soubeyrand commented:
@edsiper ping ;-)

lrg-ev0 commented:
Any timeline on this feature being released?  If I am reading correctly this will allow for setting a specific parser for each container within a pod, as well as setting the exclude option?

edsiper commented:
@yann-soubeyrand just put a comment on #1420

yann-soubeyrand commented:
@edsiper following the explanations in #1420, here is what I did in this PR:
- I refactored almost all the tests: some of the tests were more or less testing the same thing, some were not testing something particular, some didn't test the functionality correctly, and a lot of tests were missing.
- Doing so, it uncovered a lot of bugs: some of them have been fixed separately (like the Json escaping issue), and I fixed others to make the tests pass. This is the case for a bug which gives incorrect timestamps on records when no time can be correctly parsed (the timestamp put by Docker must be kept in this case) and for the parser annotation handling which behaved incorrectly.

At the beginning, I was just trying to add the ability to have per container and per stream exclude annotations. I had really hard work making it work because I based my work on the code of the parser annotation, assuming it was working correctly because it had tests. Then I realized that the tests were clearly insufficient and that the functionality didn't work as intended in a lot of cases. This is why I started this refactoring work, which took me a whole lot of time because new bugs appeared as I added tests and I want to have a PR with all tests passing.

So, please, even if it's a really hard PR to review, please take the time to do so, I strongly believe it will benefit Fluent Bit, the number of bug fixed because of it being a proof of that matter. Also, I've work based on this PR waiting (notably the exclude annotation which you merged the documentation for ;-) ).

edsiper commented:
@yann-soubeyrand 

thanks for the explanations. I am reviewing them now. 

I appreciate the effort and time you are investing on this. /me reviewing #1420

edsiper commented:
we are ready, please rebase this one.

yann-soubeyrand commented:
> we are ready, please rebase this one.

Done ;-)

yann-soubeyrand commented:
@edsiper ping ;-)

edsiper commented on src/flb_parser_json.c:
```diff
@@ -165,7 +165,7 @@ int flb_parser_json_do(struct flb_parser *parser,
         flb_warn("[parser:%s] Invalid time format %s for '%s'.",
                  parser->name, parser->time_fmt, tmp);
-        time_lookup = time(NULL);
+        time_lookup = 0;
```
what are the side effects of this change for a non-kubernetes use case ?

yann-soubeyrand replied:
I didn't find side effects when checking the code. Anyway, falling back to current time should not be the responsibility of the parser as there is no way to distinguish a fallback from a successful time parsing. This should be the responsibility of the calling code.

edsiper replied:
looking at the code flb_parser_regex, it's always assumed if time lookup resolution fails, assign the current timestamp. Let's preserve that behavior.

edsiper commented:
just submitted a comment

edsiper commented:
thanks!
